### PR TITLE
Plugin and lib configure

### PR DESCRIPTION
**Pull request by MathiasPaulin**
This PR improve the usability of the Radium ecosystem in Client project.
It brings several changes in the cmake configuration procedure of Radium and its client components. 
These changes allow to configure all the Radium components (libraries, plugins, applications) using only a small number of function call that ensure coherent configuration, compile, and install of the components.
This allow also to develop client components (applications, plugins and libraries) based on the Radium ecosystem in a coherent and efficient way.
   - Client libraries : configuration script allow to define Radium client libraries, with or without internal resources, so that one can install and use these libraries anywhere but in the way internal Radium components are used (Core, Engine, ...)
   - Plugins : development of plugins that define only the interface between any Plugin Compatible Radium application and any service providing library (based on Radium or not) with correct installation
   - Application : development of bundled applications with plugins and libraries taken outside of the Radium ecosystem.

The repositories  Radium-Apps and Radium-PluginExample will be updated to follow the new cmake configuration.
The (storm private) repository https://gitlab.com/Storm-IRIT/radium-private-plugins/Radium-glTFPlugin.git use the Radium distribution of this PR and gives an exemple on how to develop a Radium-based library and use it into a Radium plugin or a independent command-line application (with or without installation).

* **Please check if the PR fulfills these requirements**
- [x] The commit message follows our guidelines
- [ ] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)

Be aware that the PR request cannot be accepted if it doesn't pass the Continuous Integration tests.

* **What kind of change does this PR introduce?** (Bug fix, feature, docs update, ...)
Improved cmake scripts to configure Libraries, Plugins and Application development.

* **What is the current behavior?** (You can also link to an open issue here)
There are several cmake functions doing roughly the same job but preventing an easy usage in client project. Only client applications or client plugins could be developed, there is a lack of configuration functions for client libraries.

* **What is the new behavior (if this is a feature change)?**
Client can configure Libraries, plugins, and application the same way these components are configured in the Radium internals.
This improve the usability of the Radium ecosystem into client project.

* **Does this PR introduce a breaking change?** (What changes might users need to make in their application due to this PR?)
No

* **Other information**:

### CODE REVIEW

**Comment by nmellado** [comment]
Thanks @MathiasPaulin for this PR.
We need to test it on windows, unfortunately my configuration is broken.
I will try to add this check on Radium-Releases to validate it ASAP.

**Comment by dlyr** [comment]
Can you add a dummy use case with two plugins, the first one define a lib, the second one use this lib. (added as an independent project in the tests directory, we will hand tested for the moment)

**Comment by nmellado** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Clones added
============
- tests/PluginsAndLib/PluginA_WithHelperLib/Plugin/DummyPlugin.hpp  1
- tests/PluginsAndLib/PluginB_UsingExternalHelperLib/Plugin/DummyParasitePlugin.hpp  1
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/STORM/Radium-Engine/pullRequest?prid=5459634&bid=17894679)

**Comment by nmellado** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Clones added
============
- tests/PluginsAndLib/PluginA_WithHelperLib/Plugin/DummyPlugin.hpp  1
- tests/PluginsAndLib/PluginB_UsingExternalHelperLib/Plugin/DummyParasitePlugin.hpp  1
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/STORM/Radium-Engine/pullRequest?prid=5459634&bid=17894679)

**Comment by nmellado** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Clones added
============
- tests/PluginsAndLib/PluginB/Plugin/PluginB.hpp  1
- tests/PluginsAndLib/PluginA/Plugin/PluginA.hpp  1
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/STORM/Radium-Engine/pullRequest?prid=5459634&bid=17894679)

**Comment by nmellado** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Clones added
============
- tests/PluginsAndLib/PluginB/Plugin/PluginB.hpp  1
- tests/PluginsAndLib/PluginA/Plugin/PluginA.hpp  1
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/STORM/Radium-Engine/pullRequest?prid=5459634&bid=17894679)

**Comment by MathiasPaulin** [comment]
> I think if you stick with PluginA and PluginB it's fine,

Done

> Also, as I have started to do, but since I'm not sure I haven't finished, I think only one README.md is sufficient, in PluginsAndLib. This README should describe the configuration and compil procedure for both plugins.

Well, as you can take any plugin, or the library and configure/compile/install/use without the need of others, I think a simple README is welcome in each base directory.

**Comment by nmellado** [comment]
I tried to use compile PluginA, however the following projects do not compile:
 - https://github.com/MathiasPaulin/Radium-Engine/tree/pluginAndLibConfigure/tests/PluginsAndLib/PluginA/
```
 Configure library ExampleLibrary for insertion into Radium exosystem at /tmp/QtCreator-SBq72V/qtc-cmake-XX1AIVS0/installed-GNU
 [configure_plugin] Request to install the helper library ExampleLibrary::ExampleLibrary for plugin PluginA
Installing ExampleLibrary
CMake Error at XXXXl/Radium-Engine-installer/lib/cmake/Radium/radium_setup_functions.cmake:468 (install):
  install TARGETS given target "ExampleLibrary" which does not exist in this
  directory.
Call Stack (most recent call first):
  Plugin/CMakeLists.txt:44 (configure_radium_plugin)
```
 - https://github.com/MathiasPaulin/Radium-Engine/tree/pluginAndLibConfigure/tests/PluginsAndLib/PluginA/Plugin
```
 [configure_plugin] Request to install the helper library ExampleLibrary for plugin PluginA
CMake Error at XXXXl/Radium-Engine-installer/lib/cmake/Radium/radium_setup_functions.cmake:440 (get_target_property):
  get_target_property() called with non-existent target "ExampleLibrary".
Call Stack (most recent call first):
  CMakeLists.txt:65 (configure_radium_plugin)
```

**Comment by nmellado** [comment]
I think I have found the problem: https://stackoverflow.com/a/34444949
Will try to update cmake and confirm it.

**Comment by nmellado** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Clones added
============
- tests/PluginsAndLib/PluginA/Plugin/PluginA.hpp  1
- tests/PluginsAndLib/PluginB/Plugin/PluginB.hpp  1
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/STORM/Radium-Engine/pullRequest?prid=5459634&bid=17894679)

**Comment by nmellado** [comment]
> I think I have found the problem: https://stackoverflow.com/a/34444949
> Will try to update cmake and confirm it.

I confirm using more recent version of cmake fixes the problem.
I have added the minimal version in the cmake files.

**Next step** (will do it): add the plugins to the test suite, then it is all fine for me.

**Comment by nmellado** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Clones added
============
- tests/PluginsAndLib/PluginB/Plugin/PluginB.hpp  1
- tests/PluginsAndLib/PluginA/Plugin/PluginA.hpp  1
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/STORM/Radium-Engine/pullRequest?prid=5459634&bid=17894679)

**Comment by nmellado** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Clones added
============
- tests/PluginsAndLib/PluginB/Plugin/PluginB.hpp  1
- tests/PluginsAndLib/PluginA/Plugin/PluginA.hpp  1
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/STORM/Radium-Engine/pullRequest?prid=5459634&bid=17894679)

**Comment by nmellado** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Clones added
============
- tests/PluginsAndLib/PluginB/Plugin/PluginB.hpp  1
- tests/PluginsAndLib/PluginA/Plugin/PluginA.hpp  1
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/STORM/Radium-Engine/pullRequest?prid=5459634&bid=17894679)

**Comment by nmellado** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Clones added
============
- tests/PluginsAndLib/PluginB/Plugin/PluginB.hpp  1
- tests/PluginsAndLib/PluginA/Plugin/PluginA.hpp  1
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/STORM/Radium-Engine/pullRequest?prid=5459634&bid=17894679)

**Comment by nmellado** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Clones added
============
- tests/ExamplePluginWithLib/Downstream/Plugin/PluginB.hpp  1
- tests/ExamplePluginWithLib/Upstream/Plugin/PluginA.hpp  1
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/STORM/Radium-Engine/pullRequest?prid=5459634&bid=17894679)

File: tests/ExamplePluginWithLib/Upstream/README.md [hunk 1]
```diff
@@ -0,0 +1,23 @@
+# Plugin sharing library - Downstream
+This example illustrates how to ship a library with a Radium plugin, and make the library available
+to other plugins.
+
+## Configure and compile
+```bash
+mkdir build
+ce build
```
Comment (dlyr) [inline tests/ExamplePluginWithLib/Upstream/README.md:8]:
```suggestion
cd build
```

File: tests/ExamplePluginWithLib/Upstream/README.md [hunk 2]
```diff
@@ -0,0 +1,23 @@
...
+cmake -DRadium_DIR=/pathToInstalledRadium/lib/cmake/Radium/  -DQt5_DIR=/pathToInstalledQt5/lib/cmake/Qt5 ..`
```
Comment (dlyr) [inline tests/ExamplePluginWithLib/Upstream/README.md:9]:
```suggestion
cmake -DRadium_DIR=/pathToInstalledRadium/lib/cmake/Radium/..`
```

File: tests/ExamplePluginWithLib/Downstream/README.md [hunk 3]
```diff
@@ -0,0 +1,18 @@
+# Plugin sharing library - Downstream
+This example illustrates how to use a library shipped from another plugin (should be installed first).
+
+## Configure
+`cmake -DRadium_DIR=/pathToInstalledRadium/lib/cmake/Radium/  -DQt5_DIR=/pathToInstalledQt5/lib/cmake/Qt5
```
Comment (dlyr) [inline tests/ExamplePluginWithLib/Downstream/README.md:5]: Do Qt5 installation handled specially for this plugin ?
I don't think we need to specify that here.
```suggestion
`cmake -DRadium_DIR=/pathToInstalledRadium/lib/cmake/Radium/
```

File: tests/ExamplePluginWithLib/Downstream/README.md [hunk 4]
```diff
@@ -0,0 +1,18 @@
...
+-DDummyLibrary=/pathToInstalledDummyLibrary ..`
```
Comment (dlyr) [inline tests/ExamplePluginWithLib/Downstream/README.md:6]: What is DummyLibrary ? Upstream ?
```suggestion
-DDummyLibrary=/pathToInstalledUpstreamLibrary
Library ..`
```

File: tests/ExamplePluginWithLib/Downstream/Plugin/CMakeLists.txt [hunk 5]
```diff
@@ -0,0 +1,41 @@
+
+# This Plugin uses an external helper library (DummyLibrary).
```
Comment (dlyr) [inline tests/ExamplePluginWithLib/Downstream/Plugin/CMakeLists.txt:2]:
```suggestion
# This Plugin uses an external helper library (ExampleLibraryUpstream).
```

File: tests/ExamplePluginWithLib/Downstream/Plugin/CMakeLists.txt [hunk 6]
```diff
@@ -0,0 +1,41 @@
...
+# This library can be searched and used in its build tree or in an installed one.
+if (NOT TARGET ExampleLibraryUpstream)
+    message(STATUS "Searching for package ExampleLibrary")
+    find_package(ExampleLibraryUpstream REQUIRED)
+endif ()
+#once found (or used from the same buildtree than this project, the helper library has name DummyLibrary::DummyLibrary
```
Comment (dlyr) [inline tests/ExamplePluginWithLib/Downstream/Plugin/CMakeLists.txt:8]:
```suggestion
#once found (or used from the same buildtree than this project, the helper library has name ExampleLibraryUpstream::ExampleLibraryUpstream
```

**Comment by nmellado** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Clones added
============
- tests/ExamplePluginWithLib/Upstream/Plugin/PluginA.hpp  1
- tests/ExamplePluginWithLib/Downstream/Plugin/PluginB.hpp  1
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/STORM/Radium-Engine/pullRequest?prid=5459634&bid=17894679)

**Comment by nmellado** [comment]
Thanks @dlyr for the comments.
I am still working on the cmake files to compile this library for testing.

**Comment by nmellado** [comment]
Ready for reviews

**Comment by nmellado** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Clones added
============
- tests/ExamplePluginWithLib/Downstream/Plugin/PluginB.hpp  1
- tests/ExamplePluginWithLib/Upstream/Plugin/PluginA.hpp  1
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/STORM/Radium-Engine/pullRequest?prid=5459634&bid=17894679)

**Comment by nmellado** [comment]
@dlyr  sorry I didn't fix the typos, can you check the whole process, then I'll fix this when it is all good.

**Comment by nmellado** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Clones added
============
- tests/ExamplePluginWithLib/Upstream/Plugin/PluginA.hpp  1
- tests/ExamplePluginWithLib/Downstream/Plugin/PluginB.hpp  1
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/STORM/Radium-Engine/pullRequest?prid=5459634&bid=17894679)

**Comment by nmellado** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Clones added
============
- tests/ExamplePluginWithLib/Upstream/Plugin/PluginA.hpp  1
- tests/ExamplePluginWithLib/Downstream/Plugin/PluginB.hpp  1
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/STORM/Radium-Engine/pullRequest?prid=5459634&bid=17894679)

**Comment by nmellado** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Clones added
============
- tests/ExamplePluginWithLib/Upstream/Plugin/PluginA.hpp  1
- tests/ExamplePluginWithLib/Downstream/Plugin/PluginB.hpp  1
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/STORM/Radium-Engine/pullRequest?prid=5459634&bid=17894679)

File: tests/ExamplePluginWithLib/Upstream/README.md (hunk 2 above)
Comment (MathiasPaulin) [reply tests/ExamplePluginWithLib/Upstream/README.md:8]: As the plugin search for the package Qt, this must be given to cmake if Qt5 is not installed on "standard" location. This is mostly (and quite always) the case on Windows or MacOs where the user might install Qt5 in user space, or system-wide, in whatever directory `/pathToInstalledQt5` he wants.

File: tests/ExamplePluginWithLib/Downstream/README.md (hunk 3 above)
Comment (MathiasPaulin) [reply tests/ExamplePluginWithLib/Downstream/README.md:5]: see my previous comment. But this will be explained differently.

**Comment by nmellado** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Clones added
============
- tests/ExamplePluginWithLib/Upstream/Plugin/PluginA.hpp  1
- tests/ExamplePluginWithLib/Downstream/Plugin/PluginB.hpp  1
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/STORM/Radium-Engine/pullRequest?prid=5459634&bid=17894679)

File: tests/ExamplePluginWithLib/CMakeLists.txt [hunk 7]
```diff
@@ -45,10 +45,11 @@ macro(addLocalExternalFolder NAME FOLDER )
...
                 -DRadium_DIR=${Radium_DIR}
-                -DCMAKE_CXX_FLAGS=${CMAKE_CXX_FLAGS}
-                -DCMAKE_CXX_COMPILER=${CMAKE_CXX_COMPILER}
-                -DCMAKE_CC_COMPILER=${CMAKE_CC_COMPILER}
-                -DCMAKE_BUILD_TYPE=${CMAKE_BUILD_TYPE}
+        -DQt5_DIR=${Qt5_DIR}
+        -DCMAKE_CXX_FLAGS=${CMAKE_CXX_FLAGS}
+        -DCMAKE_CXX_COMPILER=${CMAKE_CXX_COMPILER}
+        -DCMAKE_C_COMPILER=${CMAKE_C_COMPILER}
+        -DCMAKE_BUILD_TYPE=${CMAKE_BUILD_TYPE}
```
Comment (nmellado) [inline tests/ExamplePluginWithLib/CMakeLists.txt:52]: Don't change the indentation..
```suggestion
                -DQt5_DIR=${Qt5_DIR}
                -DCMAKE_CXX_FLAGS=${CMAKE_CXX_FLAGS}
                -DCMAKE_CXX_COMPILER=${CMAKE_CXX_COMPILER}
                -DCMAKE_C_COMPILER=${CMAKE_C_COMPILER}
                -DCMAKE_BUILD_TYPE=${CMAKE_BUILD_TYPE}
```

Comment (MathiasPaulin) [reply tests/ExamplePluginWithLib/CMakeLists.txt:52]: formatted according to clang-format !

Comment (nmellado) [reply tests/ExamplePluginWithLib/CMakeLists.txt]: So apply it on the surrounding lines as well

**Comment by nmellado** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Clones added
============
- tests/ExamplePluginWithLib/Upstream/Plugin/PluginA.hpp  1
- tests/ExamplePluginWithLib/Downstream/Plugin/PluginB.hpp  1
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/STORM/Radium-Engine/pullRequest?prid=5459634&bid=17894679)

**Comment by nmellado** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Clones added
============
- tests/ExamplePluginWithLib/Upstream/Plugin/PluginA.hpp  1
- tests/ExamplePluginWithLib/Downstream/Plugin/PluginB.hpp  1
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/STORM/Radium-Engine/pullRequest?prid=5459634&bid=17894679)

**Comment by nmellado** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Clones added
============
- tests/ExamplePluginWithLib/Downstream/Plugin/PluginB.hpp  1
- tests/ExamplePluginWithLib/Upstream/Plugin/PluginA.hpp  1
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/STORM/Radium-Engine/pullRequest?prid=5459634&bid=17894679)

**Comment by nmellado** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Clones added
============
- tests/ExamplePluginWithLib/Downstream/Plugin/PluginB.hpp  1
- tests/ExamplePluginWithLib/Upstream/Plugin/PluginA.hpp  1
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/STORM/Radium-Engine/pullRequest?prid=5459634&bid=17894679)

**Pull request merged**